subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -571,100 +590,194 @@ static void zoap_options_to_path(struct zoap_option *opt, int options_count,
+{
+	size_t i;
+
+	if (!pending || !messages) {
+		return NULL;
+	}
+
+	for (i = 0; i < len; i++) {
+		if (messages[i].ctx && messages[i].pending == pending) {
+			return &messages[i];
+		}
+	}
+
+	return NULL;
+}
+
+struct lwm2m_message *lwm2m_get_message(struct lwm2m_ctx *client_ctx)
+{
+	size_t i;
+
+	for (i = 0; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES; i++) {
+		if (!client_ctx->messages[i].ctx) {
+			client_ctx->messages[i].ctx = client_ctx;
+			return &client_ctx->messages[i];
+		}
+	}
+
+	return NULL;
+}
+
+void lwm2m_release_message(struct lwm2m_message *msg)
+{
+	if (!msg) {
+		return;
+	}
+
+	if (msg->pending) {
+		zoap_pending_clear(msg->pending);
+		msg->pending = NULL;
+	}
+
+	if (msg->reply) {
+		/* make sure we want to clear the reply */
+		zoap_reply_clear(msg->reply);
+		msg->reply = NULL;
+	}
+
+	memset(msg, 0, sizeof(*msg));
+}
+
+int lwm2m_init_message(struct lwm2m_message *msg)
 {
+	struct net_pkt *pkt;
+	struct net_app_ctx *app_ctx;
 	struct net_buf *frag;
 	int r;
 
-	*pkt = net_app_get_net_pkt(app_ctx, AF_UNSPEC, BUF_ALLOC_TIMEOUT);
-	if (!*pkt) {
+	if (!msg || !msg->ctx) {
+		SYS_LOG_ERR("LwM2M message is invalid.");
+		return -EINVAL;
+	}
+
+	app_ctx = &msg->ctx->net_app_ctx;
+	pkt = net_app_get_net_pkt(app_ctx, AF_UNSPEC, BUF_ALLOC_TIMEOUT);
+	if (!pkt) {
 		SYS_LOG_ERR("Unable to get TX packet, not enough memory.");
 		return -ENOMEM;
 	}
 
-	frag = net_app_get_net_buf(app_ctx, *pkt,
+	frag = net_app_get_net_buf(app_ctx, pkt,
 				   BUF_ALLOC_TIMEOUT);
 	if (!frag) {
 		SYS_LOG_ERR("Unable to get DATA buffer, not enough memory.");
-		net_pkt_unref(*pkt);
-		*pkt = NULL;
-		return -ENOMEM;
+		r = -ENOMEM;
+		goto cleanup;
 	}
 
-	r = zoap_packet_init(zpkt, *pkt);
+	r = zoap_packet_init(&msg->zpkt, pkt);
 	if (r < 0) {
 		SYS_LOG_ERR("zoap packet init error (err:%d)", r);
-		return r;
+		goto cleanup;
 	}
 
 	/* FIXME: Could be that zoap_packet_init() sets some defaults */
-	zoap_header_set_version(zpkt, 1);
-	zoap_header_set_type(zpkt, type);
-	zoap_header_set_code(zpkt, code);
+	zoap_header_set_version(&msg->zpkt, 1);
+	zoap_header_set_type(&msg->zpkt, msg->type);
+	zoap_header_set_code(&msg->zpkt, msg->code);
 
-	if (mid > 0) {
-		zoap_header_set_id(zpkt, mid);
+	if (msg->mid > 0) {
+		zoap_header_set_id(&msg->zpkt, msg->mid);
 	} else {
-		zoap_header_set_id(zpkt, zoap_next_id());
+		zoap_header_set_id(&msg->zpkt, zoap_next_id());
 	}
 
-	/* tkl == 0 is for a new TOKEN, tkl == -1 means dont set */
-	if (token && tkl > 0) {
-		zoap_header_set_token(zpkt, token, tkl);
-	} else if (tkl == 0) {
-		zoap_header_set_token(zpkt, zoap_next_token(), 8);
+	/*
+	 * tkl == 0 is for a new TOKEN
+	 * tkl == LWM2M_MSG_TOKEN_LEN_SKIP means dont set
+	 */
+	if (msg->tkl == 0) {
+		zoap_header_set_token(&msg->zpkt, zoap_next_token(),
+				      MAX_TOKEN_LEN);
+	} else if (msg->token && msg->tkl != LWM2M_MSG_TOKEN_LEN_SKIP) {
+		zoap_header_set_token(&msg->zpkt, msg->token, msg->tkl);
+	}
+
+	if (msg->type == ZOAP_TYPE_CON) {
```

Comment:
taste issue, but you could test the contrary here and return 0 if so.

Then code below would not be as indented as it is.